### DISCUSSION TITLE
how to go about debugging topLevelAwait hangs?

### DISCUSSION THREAD

**Discussion by starpit** [webpack/webpack#17018] [General]
we have begun to use [Ink](https://github.com/vadimdemedes/ink) which, on its own, and via its dependence on [Yoga](https://github.com/facebook/yoga), leverage topLevelAwait.

we are seeing odd and hard to reproduce hangs that show up only after the switch from ink 3 (that doesn't use topLevelAwait) and ink 4 (that does...). 

it's hard for us to tell where the race is (us? ink? yoga? webpack?). how do the webpack developers go about debugging this kind of thing?

thanks,
@starpit

**Comment by alexander-akait** [comment]
Can you try to debug code using node debbuger?

**Reply by starpit** [reply]
nice idea! when i connect chrome devtools and hit Pause... devtools hangs without any feedback. i'll keep investigating.

fwiw, this is on a production build.

**Reply by starpit** [reply]
the hang seems to happen in code that does a dynamic import of code that does a static import of code that does a top-level await. it's the dynamic import -> top-level await that seems (so far) to be the cause of the hang. the app just hangs on that dynamic import, never resolving.

and this only seems to be a problem with production builds.

**Reply by alexander-akait** [reply]
Do you have circular dependencies?

**Reply by starpit** [reply]
hmm, i don't think so. we are using typescript, and it is happy. the bundled app seems fine in dev mode.

the only thing i have been able to determine so far is that `yoga-wasm-web/auto` (imported via `ink`) does a dynamic import of a wasm file; and it does this as a top-level await.

and our webpack config had no rules for this file type.  so... presumably that dynamic import failed, causing the top-level await to ...? fail as an unhandled promise rejection? which somehow resulted in a silent hang?

and why this sometimes works? why this works in development mode always? why would the app (a node target, not browser) silently hang when loading these bundles, rather than report the issue? i'm not sure.

we've added a wasm module.rule and will continue to monitor the situation.

**Reply by alexander-akait** [reply]
> and why this sometimes works? why this works in development mode always? why would the app (a node target, not browser) silently hang when loading these bundles, rather than report the issue? i'm not sure.

A lot of questions :smile: need debug, if you have problems with it ping me I will look at repo
